hybrisprovider.cpp

```diff
@@ -593,74 +256,41 @@ HybrisProvider::HybrisProvider(QObject *parent)
         }
     }
 
-    const hw_module_t *hwModule;
-
-    int error = hw_get_module(GPS_HARDWARE_MODULE_ID, &hwModule);
-    if (error) {
-        qWarning("Android GPS interface not found, error %d\n", error);
-        return;
-    }
-
-    qWarning("Android GPS hardware module \"%s\" \"%s\" %u.%u\n", hwModule->id, hwModule->name,
-             hwModule->module_api_version, hwModule->hal_api_version);
-
-    error = hwModule->methods->open(hwModule, GPS_HARDWARE_MODULE_ID,
-                                    reinterpret_cast<hw_device_t **>(&m_gpsDevice));
-    if (error) {
-        qWarning("Failed to open GPS device, error %d\n", error);
-        return;
+    QDir pluginsDir(qApp->applicationDirPath());
+    bool binderBackendFound = false;
+    foreach (QString fileName, pluginsDir.entryList(QDir::Files)) {
+        if (fileName.contains("libgeoclue-hybris-binder.so")) {
```

Comment:
Could plugin selection and loading be slightly more generic? Like, check all libraries with certain prefix and pick the first one that implements the required interface, or something like that. Do we really need to support scenario with multiple plugins installed?

Reply:
We always have the hal plugin present so we want to prefer the binder plugin. If we try to use the first one found then we would have to try to initialize the plugin and if initializing fails then continue to the next one found. This is possible if that is wanted.

Reply:
Ok, so multiple plugins is a realistic and apparently meaningful scenario.

If it's hal plugin that for whatever reason has to be always present (sounds strange to me but suppose it does make sense) then I suggest that the logic should be slightly different - look for any plugin other than hal, if it's found then use that, otherwise use the default hal thing.

In other words, hardcode as few file names as possible (only `libgeoclue-hybris-hal.so`).

Reply:
Presence of HAL plugin is for backwards compatibility to make sure building only the default package would be enough in old adaptations. Other option would be to add some suitable dependencies in each package so by default the HAL package would be used but if some other package is forcefully installed then the HAL plugin package would not be installed at all and we would have one plugin installed at all times.

Reply:
Yes, that's kind of like what I had in mind. Provider's package may require something (e.g. `geoclue-providers-backend`) and backend packages provide it. Then it's a matter of which backend package gets pulled in by the build. That would naturally make backends exclusive, and we won't have to bother with multiple backends being installed simultaneously. I'm not quite sure though, how to make it most convenient for porters (i.e. how to automatically provide a reasonable default without any additional work being done), that's something you probably know better than I.